aws_lambda_builders/workflows/rust_cargo/cargo_lambda.py

```diff
@@ -0,0 +1,102 @@
+            with the prepended additional search paths
+
+        :type executable_search_paths: list
+        :param executable_search_paths: List of paths to the NPM package binary utilities. This will
+            be used to find embedded esbuild at runtime if present in the package
+
+        :type osutils: aws_lambda_builders.workflows.rust_cargo.actions.OSUtils
+        :param osutils: An instance of OS Utilities for file manipulation
+        """
+        self._which = which
+        self._executable_search_paths = executable_search_paths
+        self._osutils = osutils
+
+    def check_cargo_lambda_installation(self):
+        """
+        Checks if Cargo Lambda is in the system
+
+        :raises aws_lambda_builders.workflows.rust_cargo.exceptions.CargoLambdaExecutionException:
+            when Cargo Lambda is not installed in the system to run the command.
+        """
+
+        LOG.debug("checking for cargo-lambda")
+        binaries = self._which("cargo-lambda", executable_search_paths=self._executable_search_paths)
+        LOG.debug("potential cargo-lambda binaries: %s", binaries)
+
+        if binaries:
+            return binaries[0]
+        else:
+            raise CargoLambdaExecutionException(
+                message="Cannot find Cargo Lambda. "
+                "Cargo Lambda must be installed on the host machine to use this feature. "
+                "Follow the gettings started guide to learn how to install it: "
+                "https://www.cargo-lambda.info/guide/getting-started.html"
+            )
+
+    def run(self, command, cwd):
+        """
+        Runs the build command.
+
+        :type command: str
+        :param args: Cargo Lambda command to run
+
+        :type cwd: str
+        :param cwd: Directory where to execute the command (defaults to current dir)
+
+        :rtype: str
+        :return: text of the standard output from the command
+
+        :raises aws_lambda_builders.workflows.rust_cargo.exceptions.CargoLambdaExecutionException:
+            when Cargo Lambda is not installed in the system to run the command.
+
+        :raises aws_lambda_builders.workflows.rust_cargo.exceptions.CargoLambdaExecutionException:
+            when the command executes with a non-zero return code. The exception will
+            contain the text of the standard error output from the command.
+        """
+
+        self.check_cargo_lambda_installation()
+
+        LOG.debug("Executing cargo-lambda: %s", " ".join(command))
+        if LOG.isEnabledFor(logging.DEBUG):
+            if "RUST_LOG" not in os.environ:
+                os.environ["RUST_LOG"] = "debug"
+            LOG.debug("RUST_LOG environment variable set to `%s`", os.environ.get("RUST_LOG"))
+        cargo_process = self._osutils.popen(
+            command,
+            stderr=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            cwd=cwd,
+        )
+        out, err = cargo_process.communicate()
```

Comment:
While this works, customer will see the process hangs for a while before the entire `cargo lambda` build log is printed. Can we stream the output instead?

We can replace L94-L102 with:
```
        stdout = ""
        # Create a buffer and use a thread to gather the stderr stream into the buffer
        stderr_buf = io.BytesIO()
        stderr_thread = threading.Thread(target=shutil.copyfileobj, args=(cargo_process.stderr, stderr_buf), daemon=True)
        stderr_thread.start()

        # Log every stdout line by iterating
        for line in cargo_process.stdout:
            decoded_line = line.decode("utf-8").strip()
            LOG.info(decoded_line)
            # Gather total stdout
            stdout += decoded_line

        # Wait for the process to exit and stderr thread to end.
        return_code = cargo_process.wait()
        stderr_thread.join()

        if return_code != 0:
            # Raise an Error with the appropriate value from the stderr buffer.
            raise CargoLambdaExecutionException(message=stderr_buf.getvalue().decode("utf8").strip())
        return stdout
```